File: eventmachine.js
'use strict';

/* A Toy Event Machine 
@author - Ankur Anand
Tested Using Nodejs 6.2*/

// Event
class Event	{
	// Structure of Event Object
	constructor (etype, data){
		this.etype = etype; // event type
		this.data = data; // data associated with the event
	}

	toString() {
		// toString representation of object
		return `(${this.etype}, ${this.data})`;
	}
}

// Event Handler
function eventHandler_A (event) {
	// Handle Event of type A
	if 	( ! event instanceof Event ){
		console.log('Not an Event Object');
	}
	else {
		console.log(`Event of type ${event.etype} occurred with Data 
			${event.data}`);
	}
	
}

function eventHandler_B(event) {
	// Handle Event of type B
	if ( ! event instanceof Event ){
		console.log('Not an Event Object');
	}
	else {
		console.log(`Event of type ${event.etype} occurred with Data 
			${event.data.toUpperCase()}`);
	}
}


class Queue {
	// A queue class for storing Events FIFO
	constructor (){
		this.items = [];
	}

	// methods acting on the queue
	isEmpty() {
		return (this.items).length === 0;
	}

	enqueue(item) {
		// push at the beginning of the array
		this.items.unshift(item);
	}

	dequeue(item) {
		// pop the last element
		return this.items.pop();
	}

	size() {
		return (this.items).length;
	}
	
}


let events = new Queue();
events.enqueue(new Event('A','Hello'));
events.enqueue(new Event('B', 'Event-Driven'));
events.enqueue(new Event('A', 'World'));


let handler_dispatcher = {
	'A' : eventHandler_A,
	'B' : eventHandler_B
}
// Event loop
while (! events.isEmpty()) {
	let e = events.dequeue();
	/*if ( e.etype === 'A') {
		eventHandler_A(e);
	}
	if ( e.etype === 'B') {
		eventHandler_B(e);
	}*/

	handler_dispatcher[e.etype](e);
}

File: eventmachine.py
#################################
### A Toy Event machine 	#####
### @author - Ankur Anand	#####
### Tested using python 3.5.2 ###
#################################


### Event
class Event(object):
	""" Structure for event object"""
	def __init__(self, etype, data):
		self.etype = etype # type of the event
		self.data = data # data associated with the type events

	def __str__(self):
		""" String representation of the class instance """
		return 'Event({etype}, {data})'.format(
				etype = self.etype,
				data = repr(self.data)
			)
	def __repr__(self):
		return self.__str__()


####### Event Handlers. In our toy implementation we are defining two event
####### handlers A and B

def eventHandler_A(event):
	""" receives a event of type A and print it"""
	if not type(event) is Event:
		print("Not an Event")
	print("Event Type A Occurred with data : ",event.data)

def eventHandler_B(event):
	""" receives a event of type B and uppercase it data"""
	if not type(event) is Event:
		print("Not an Event")
	print("Event Type B Occurred with data : ", (event.data).upper()) 



########### Implementing a Queue to store events ###

class Queue(object):
	""" A FIFO Queue"""
	def __init__(self):
		self.items = []

	def isEmpty(self):
		""" return True if the Queue is empty"""
		return self.items == []

	def enqueue(self, item):
		""" Add the element at the beginning of Queue """
		self.items.insert(0, item)

	def dequeue(self):
		""" Remove the element from Queue from the last"""
		return self.items.pop()

	def size(self):
		""" Return the size of the Queue"""
		return len(self.items)



events = Queue()
events.enqueue(Event('A','Hello'))
events.enqueue(Event('B','event-driven'))
events.enqueue(Event('A','world'))

### Event Loop
while not events.isEmpty():
	e = events.dequeue()
	if e.etype == 'A':
		eventHandler_A(e)
	if e.etype == 'B':
		eventHandler_B(e)


##### OutPut ######
"""
Event Type A Occurred with data :  Hello
Event Type B Occurred with data :  EVENT-DRIVEN
Event Type A Occurred with data :  world
"""
